### PR TITLE
ISPN-13880 Create an event logger listener

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-13880

* EventLogger now extends Listenable, so listeners can subscribe to
receive notifications about logged data.

* Created the `@Logged` annotation, to use in the logger listeners.

Then I'll update #10092 to use the listener.

### CODE REVIEW

File: server/runtime/src/main/java/org/infinispan/server/logging/events/impl/EventLoggerNotifierImpl.java
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/logging/events/impl/EventLoggerNotifierImpl.java:31]: Capacity should be 1.

File: server/runtime/src/main/java/org/infinispan/server/logging/annotation/Logged.java
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/logging/annotation/Logged.java:23]: Should this be `LoggedEvent` to match the style of `CacheCreatedEvent` etc.

File: server/runtime/src/main/java/org/infinispan/server/logging/annotation/Logged.java
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/logging/annotation/Logged.java:10]: I think `EventLoggerNotifier`, and `Logged` should be moved to `core`. The `BasicEventLogger` and other `EventLogger` impls can then just call `notifyLog` as required as there's no cache to listen to.

File: server/runtime/src/main/java/org/infinispan/server/logging/events/EventLoggerNotifier.java
Comment (ryanemerson) [inline server/runtime/src/main/java/org/infinispan/server/logging/events/EventLoggerNotifier.java:16]: Just `notify`?

File: core/src/main/java/org/infinispan/util/logging/events/EventLogger.java
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/util/logging/events/EventLogger.java:23]: I don't think we need to extend the `EventLogger` to be `Listenable`, instead we can just register the `Loggable` listener to the `EmbeddedCacheManager` instance as event logs are already globally scoped.

Comment (jabolina) [reply core/src/main/java/org/infinispan/util/logging/events/EventLogger.java:23]: Do you mean we register directly into the `EmbeddedCacheManager` instead of the `EventLogger`?

Comment (ryanemerson) [reply core/src/main/java/org/infinispan/util/logging/events/EventLogger.java:23]: Exactly

Comment (jabolina) [reply core/src/main/java/org/infinispan/util/logging/events/EventLogger.java]: Moved some stuff to the `CacheManagerNotifier`, so was able to delete that notifier I've created.

**Comment by jabolina** [comment]
Updated with suggestions.

File: core/src/main/java/org/infinispan/notifications/cachemanagerlistener/CacheManagerNotifier.java
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/notifications/cachemanagerlistener/CacheManagerNotifier.java:38]: `notifyEventLogged`?

**Review by ryanemerson — COMMENTED** [review]
We should also call `notifyLogged` in `BasicEventLogger`.

File: core/src/main/java/org/infinispan/util/logging/annotation/Logged.java
Comment (jabolina) [reply core/src/main/java/org/infinispan/util/logging/annotation/Logged.java:23]: I am having a hard time with this one, sorry :anguished: The event the notifier publishes is a `LoggedEvent`, an annotation with `@EventLogged` and the argument as `LoggedEvent` could cause some confusion?

I've use the approach as the other annotations to be `@Fooed` and argument `FooedEvent`, but only `Logged` does not seem that meaningful name. If you think is ok the naming for the annotation + argument to be similar I'll update.

Comment (ryanemerson) [reply core/src/main/java/org/infinispan/util/logging/annotation/Logged.java:23]: My only issue with `Logged` on it's own is that it could cause some confusion with normal logging, but maybe that's just in my head :slightly_smiling_face:. I take your point that you're currently following an established convention, so maybe we should just keep it as `@Logged` after all.

File: core/src/main/java/org/infinispan/util/logging/events/impl/EventLogManagerImpl.java
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/util/logging/events/impl/EventLogManagerImpl.java:42]: I think `Objects#requireNonNull` is unnecessary here as the initialisation path is controlled by our CDI code. If `@Start` doesn't work as expected, then we'll have many more failures that would prevent `#getEventLogger` ever being called.

**Review by ryanemerson — APPROVED** [review]
LGTM, but I would like @wburns and/or @tristantarrant to take a look as they're much more familiar with the event/listener code base.

File: core/src/main/java/org/infinispan/util/logging/annotation/Logged.java
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/util/logging/annotation/Logged.java:23]: I'm not really enthusiastic about the introduction of this annotation, as eventually I'd like us to move away from runtime annotation usage (the new API doesn't use annotations for events anymore).

Comment (jabolina) [reply core/src/main/java/org/infinispan/util/logging/annotation/Logged.java:23]: I'll apply the same approach of the new API here and remove all the annotation related stuff

**Review by ryanemerson — COMMENTED** [review]
Overall LGTM. One minor thing would be that we might want to consider moving the `Logged` annotation to an `annotations/impl` package to make it explicit to Embedded users that this isn't for public consumption.

**Review by ryanemerson — APPROVED** [review]
I'm happy with the changes, but I'm reluctant to hit the merge button myself as I really am not an authority on this area of the code. @wburns @tristantarrant Can one of you take a look?

File: core/src/main/java/org/infinispan/util/logging/Log.java
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/util/logging/Log.java:2332]: Please do not renumber existing messages.

**Comment by tristantarrant** [comment]
One small change is needed, otherwise it LGTM.

**Comment by ryanemerson** [comment]
Thanks @jabolina

**Pull request merged**
